docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,509 @@
+should not prevent future work to allow out-of-tree provisioners.
+
+## Design
+
+We propose that:
+
+1.  Storage properties of peristent volumes is represented as labels.
+    E.g. label "aws.alpha.kubernetes.io/volume-type" with values "sc1" and "gp2"
+    represent SDD/HDD of AWS EBS. Similarly, label
+    "failure-domain.beta.kubernetes.io/region" with values like "us-east-1d"
+    represent zone where particular AWS EBS is created.
+
+    Users will request specific storage volumes by specifying
+    `claim.Spec.Selector`, such as `region NotIn [us-east-1a, us-east-1b] AND
+    volume-type == sc1`
+
+2.  A new API group called `storage` be created to hold the a `StorageClass`
+    API resource. Storage configuration is represented as `LabelSelector`,
+    specifying parameters of volumes belonging to this class.
+
+3.  `PersistentVolume.Spec.Class` attribute is added to volumes. This attribute
+    is optional and specifies which `StorageClass` instance represents
+    storage characteristics of a particular PV.
+
+    During alpha and beta development, `Class` is just an annotation and not
+    actual attribute.
+
+4.  `PersistentVolume` instances should be labelled with configuration
+    parameters that were used for their creation, such as availability zone or
+    storage type (SSD/HDD). This should be done in an admission controller.
+
+5.  `PersistentVolumeClaim.Spec.Class` attribute is added to claims. This
+    attribute specifies that only a volume with equal
+    `PersistentVolume.Spec.Class` value can satisfy a claim.
+
+    During alpha and beta development, `Class` is just an annotation and not
+    actual attribute.
+
+6.  When requesting a volume with certain configuration,
+    `PersistentVolumeClaim.Spec.Selector` should be set in a way that it matches
+    only volumes with this configuration (see point 3.)
+    When both `PersistentVolumeClaim.Spec.Selector` and
+    `PersistentVolumeClaim.Spec.Class` is set, resulting volume must match both
+    selectors in the claim and in the storage class.
+
+7.  The persistent volume controller be modified to dynamically provision
+    volumes using storage classes and claim's `Selector` to create a volume that
+    has properties requested by both `PersistentVolumeClaim.Spec.Class` and
+    `PersistentVolumeClaim.Spec.Selector`.
+
+8.  The existing provisioner plugin implementations be modified to allow
+    parameterization as appropriate via storage classes and claim selectors.
+
+9.  The existing alpha dynamic provisioning feature be phased out in the
+    next release.
+
+### Controller workflow for provisioning volumes
+
+1.  When a new claim is submitted, the controller attempts to find an existing
+    volume that will fulfill the claim.
+
+    1. If the claim has non-empty `claim.Spec.Class`, only PVs with the same
+        `pv.Spec.Class` are considered.
+
+    2. If the claim has empty `claim.Spec.Class`, all existing PVs are
+        considered.
```

Comment:
to differentiate non-specified from empty, it needs to be a pointer field
in Go.  Maybe we should do that anyway, but is "" meaning "none" really
useful?

On Thu, Jul 14, 2016 at 6:17 AM, Eric Paris <notifications@github.com>
wrote:

> In docs/proposals/volume-provisioning.md
> <https://github.com/kubernetes/kubernetes/pull/26908#discussion_r70802471>
> :
>
> > +8.  The existing provisioner plugin implementations be modified to allow
> > +    parameterization as appropriate via storage classes and claim selectors.
> > +
> > +9.  The existing alpha dynamic provisioning feature be phased out in the
> > +    next release.
> > +
> > +### Controller workflow for provisioning volumes
> > +
> > +1.  When a new claim is submitted, the controller attempts to find an existing
> > +    volume that will fulfill the claim.
> > +
> > +    1. If the claim has non-empty `claim.Spec.Class`, only PVs with the same
> > +        `pv.Spec.Class` are considered.
> > +
> > +    2. If the claim has empty `claim.Spec.Class`, all existing PVs are
> > +        considered.
>
> We all agree non-specified == any, which is the backwards compat issue.
> Isn't "" == any an uncommon kube pattern? You explicitly picked a
> StorgeClass. You picked ""
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26908/files/4e02b557bae620b6df6acdf85308d5cdc2998b25#r70802471>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFVgVOFuZEqFRoWxMpAlDR3m0qsSsSVhks5qVjblgaJpZM4IvTmn>
> .
>